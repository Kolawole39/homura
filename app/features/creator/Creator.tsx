import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { BottomBar, Step } from '../../components/SideBar/BottomBar';
import { FeedSearchBar } from '../../components/SideBar/FeedSearchBar';
import { FeedSubscribeBar } from '../../components/SideBar/FeedSubscribeBar';
import { AppDispatch } from '../../store';
import { sync } from '../source/sourceSlice';
import {
  confirmName,
  reset,
  searchUrl,
  selectCreator,
  setParserError,
  stepToEnterUrl,
} from './creatorSlice';

export const Creator: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const { step, name, refreshing, loading, parseError, link } = useSelector(
    selectCreator
  );

  const renderBottom = React.useCallback(
    (currentStep: Step) => {
      switch (currentStep) {
        case Step.EnterUrl: {
          return (
            <FeedSearchBar
              onClickError={() => dispatch(setParserError(false))}
              hasError={parseError}
              loading={loading}
              onCancel={() => dispatch(reset())}
              onSearch={(inputLink) => dispatch(searchUrl(inputLink))}
            />
          );
        }
        case Step.EnterName: {
          return (
            <FeedSubscribeBar
              link={link}
              initialName={name}
              onCancel={() => dispatch(reset())}
              onConfirm={(inputName) => dispatch(confirmName(inputName))}
            />
          );
        }
        default: {
          throw new Error('Invalid Step');
        }
      }
    },
    [dispatch, link, loading, name, parseError]
  );

  return (
    <BottomBar
      step={step}
      refreshing={refreshing}
      onClickPlus={() => dispatch(stepToEnterUrl())}
      onClickSync={() => dispatch(sync())}
      render={renderBottom}
    />
  );
};
