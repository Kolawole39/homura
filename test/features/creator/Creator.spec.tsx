import { configureStore } from '@reduxjs/toolkit';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import 'regenerator-runtime/runtime';
import { channel } from '../../../app/channel/child';
import { Step } from '../../../app/components/SideBar/BottomBar';
import { Mode } from '../../../app/constants/Mode';
import { Creator } from '../../../app/features/creator/Creator';
import * as creatorSlice from '../../../app/features/creator/creatorSlice';
import { sourceReducer } from '../../../app/features/source/sourceSlice';
import { mockStore } from '../../test-tools/mockStore';

jest.mock('../../../app/channel/child');
const mockChannel = (channel as unknown) as jest.Mocked<typeof channel>;

Enzyme.configure({ adapter: new Adapter() });

function setup(
  preloadedState: {
    creator: {
      link: string;
      name: string;
      step: Step | null;
      loading: boolean;
      parseError: boolean;
    };
    source: {
      refreshing: boolean;
    };
  } = {
    creator: {
      step: null,
      link: '',
      name: '',
      loading: false,
      parseError: false,
    },
    source: {
      refreshing: false,
    },
  }
) {
  const store = configureStore({
    reducer: { creator: creatorSlice.creatorReducer, source: sourceReducer },
    preloadedState,
  });

  const getWrapper = () =>
    mount(
      <Provider store={store}>
        <Router>
          <Creator />
        </Router>
      </Provider>
    );
  const component = getWrapper();
  return {
    store,
    component,
    buttons: component.find('IconContainer'),
  };
}

describe('Creator component', () => {
  it('should match exact snapshot', () => {
    const { store } = setup();
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Creator />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should plus button call increment', () => {
    const { buttons } = setup();
    const stepToEnterUrlSpy = jest.spyOn(creatorSlice, 'stepToEnterUrl');

    buttons.at(0).simulate('keydown');
    expect(stepToEnterUrlSpy).toBeCalled();
    stepToEnterUrlSpy.mockRestore();
  });
});

describe('Test creator actions', () => {
  it('should call searchUrl', async () => {
    const store = mockStore();
    mockChannel.checkUrl.mockResolvedValue('name');
    await store.dispatch(creatorSlice.searchUrl('fake-link'));

    expect(store.getActions()).toMatchSnapshot();
  });

  describe('should call confirmName', () => {
    it('should filter icon', async () => {
      const store = mockStore();
      mockChannel.confirm.mockResolvedValue({
        name: 'fake-name',
        sourceUrl: 'fake-url',
        posts: Array(20).fill(0),
        icon: null,
        link: 'fake-link',
        id: 20,
      });
      await store.dispatch(creatorSlice.confirmName('fake-name'));
      expect(store.getActions()).toMatchSnapshot();
    });

    it('should set count by mode', async () => {
      const store = mockStore({
        mode: Mode.Starred,
      });
      mockChannel.confirm.mockResolvedValue({
        name: 'fake-name',
        sourceUrl: 'fake-url',
        posts: Array(20).fill(0),
        icon: 'fake-icon',
        link: 'fake-link',
        id: 20,
      });
      await store.dispatch(creatorSlice.confirmName('fake-name'));
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
