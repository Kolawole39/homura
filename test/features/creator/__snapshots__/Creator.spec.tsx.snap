// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creator component should match exact snapshot 1`] = `
<div
  className="relative flex justify-between text-gray-300"
>
  <div
    className="p-2  "
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="w-5 h-5 m-auto  "
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <svg
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 6v6m0 0v6m0-6h6m-6 0H6"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
        />
      </svg>
    </div>
  </div>
  <div
    className="p-2  "
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="w-5 h-5 m-auto  "
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <svg
        className=""
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Test creator actions should call confirmName should filter icon 1`] = `
Array [
  Object {
    "payload": "fake-name",
    "type": "creator/setName",
  },
  Object {
    "payload": Object {
      "count": 20,
      "icon": undefined,
      "id": 20,
      "link": "fake-link",
      "name": "fake-name",
    },
    "type": "source/create",
  },
  Object {
    "payload": undefined,
    "type": "creator/reset",
  },
]
`;

exports[`Test creator actions should call confirmName should set count by mode 1`] = `
Array [
  Object {
    "payload": "fake-name",
    "type": "creator/setName",
  },
  Object {
    "payload": Object {
      "count": 0,
      "icon": "fake-icon",
      "id": 20,
      "link": "fake-link",
      "name": "fake-name",
    },
    "type": "source/create",
  },
  Object {
    "payload": undefined,
    "type": "creator/reset",
  },
]
`;

exports[`Test creator actions should call searchUrl 1`] = `
Array [
  Object {
    "payload": true,
    "type": "creator/setLoading",
  },
  Object {
    "payload": "fake-link",
    "type": "creator/setLink",
  },
  Object {
    "payload": "name",
    "type": "creator/setName",
  },
  Object {
    "payload": undefined,
    "type": "creator/stepToEnterName",
  },
  Object {
    "payload": false,
    "type": "creator/setLoading",
  },
]
`;
